### PR TITLE
Add extensions to get logs from mocked loggers

### PR DESCRIPTION
**Pull request by atifaziz**
## What is being addressed

Different and duplicated ways of extracting logs from mocked loggers.

## How is this addressed

Consolidated into a single way by adding extension methods that allow extracting strong-typed invocations of [`ILogger.Log<>`](https://docs.microsoft.com/en-us/dotnet/api/microsoft.extensions.logging.ilogger.log?view=dotnet-plat-ext-6.0#microsoft-extensions-logging-ilogger-log-1(microsoft-extensions-logging-loglevel-microsoft-extensions-logging-eventid-0-system-exception-system-func((-0-system-exception-system-string)))) given an instance of a mocked logger. The invocation arguments can then be used in regular assertions.

### CODE REVIEW

**Comment by bastbu** [comment]
Suggested edit:
```diff
diff --git a/Tests/Common/LoggerMockExtensions.cs b/Tests/Common/LoggerMockExtensions.cs
--- Tests/Common/LoggerMockExtensions.cs
+++ Tests/Common/LoggerMockExtensions.cs
@@ -37,10 +37,10 @@
             return from i in invocations
                    where i.Method.IsGenericMethod && i.Method.GetGenericMethodDefinition() == LogMethodInfo
                    select i.Arguments into args
                    select new LoggerLogInvocation((LogLevel)args[0],
-                                            (EventId)args[1],
-                                            (Exception?)args[3],
-                                            (string)((Delegate)args[4]).DynamicInvoke(args[2], args[3])!);
+                                                  (EventId)args[1],
+                                                  (Exception?)args[3],
+                                                  (string)((Delegate)args[4]).DynamicInvoke(args[2], args[3])!);
         }
     }
 }

```

File: Tests/Common/LoggerMockExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+    public sealed record LoggerLogInvocation(LogLevel LogLevel, EventId EventId, Exception? Exception, string Message);
+
+    public static class LoggerMockExtensions
+    {
+        private static readonly MethodInfo? LogMethodInfo = typeof(ILogger).GetMethod(nameof(ILogger.Log));
+
+        public static IEnumerable<LoggerLogInvocation> GetLogInvocations(this Mock<ILogger> mock)
+        {
+            ArgumentNullException.ThrowIfNull(mock, nameof(mock));
```
Comment (bastbu) [inline Tests/Common/LoggerMockExtensions.cs:23]: Why pass the second argument to `ArgumentNullException.ThrowIfNull`? Even when omitting it it should automatically pick up the correct parameter name, no?

**Review by bastbu — APPROVED** [review]
LGTM

**Comment by atifaziz** [comment]
> Suggested edit:
> 
> ```diff
> diff --git a/Tests/Common/LoggerMockExtensions.cs b/Tests/Common/LoggerMockExtensions.cs
> --- Tests/Common/LoggerMockExtensions.cs
> +++ Tests/Common/LoggerMockExtensions.cs
> @@ -37,10 +37,10 @@
>              return from i in invocations
>                     where i.Method.IsGenericMethod && i.Method.GetGenericMethodDefinition() == LogMethodInfo
>                     select i.Arguments into args
>                     select new LoggerLogInvocation((LogLevel)args[0],
> -                                            (EventId)args[1],
> -                                            (Exception?)args[3],
> -                                            (string)((Delegate)args[4]).DynamicInvoke(args[2], args[3])!);
> +                                                  (EventId)args[1],
> +                                                  (Exception?)args[3],
> +                                                  (string)((Delegate)args[4]).DynamicInvoke(args[2], args[3])!);
>          }
>      }
>  }
> ```

Thanks, fixed with 01e2f4050492b827d6580e1f188671bbd119e77e.

File: Tests/Common/LoggerMockExtensions.cs (hunk 1 above)
Comment (atifaziz) [reply Tests/Common/LoggerMockExtensions.cs]: Right, completely forgot it has the new `CallerArgumentExpression` applied. 🤦🏽‍♂️ Removed with 4949ae882941f222a303acf37cccc5ffac52905c.

**Pull request merged**